modyn/model_storage/internal/model_storage_manager.py

```diff
@@ -0,0 +1,288 @@
+
+        Returns:
+            dict: the plain model state derived from the model architecture of the pipeline's models.
+        """
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            model_class_name, model_config, amp = database.get_model_configuration(pipeline_id)
+        model_module = dynamic_module_import("modyn.models")
+        assert hasattr(model_module, model_class_name), f"Model {model_class_name} not available."
+
+        model_handler = getattr(model_module, model_class_name)
+        return model_handler(json.loads(model_config), "cpu", amp).model.state_dict()
+
+    def _reconstruct_model(self, model_id: int, model_state: dict, policy: ModelStoragePolicy) -> None:
+        """
+        Reconstruct the model given the model state and the model storage policy.
+        The function recursively call itself, if the model is stored as a model delta.
+        In this case it first loads the (fully stored) parent model into the model state before overwriting it
+        according to the incremental model storage policy.
+
+        Args:
+            model_id: the identifier of the model to be reconstructed.
+            model_state: the plain model state (or the loaded parent model state).
+            policy: the model storage policy containing the strategies.
+        Returns:
+            None: the model state is overwritten in order to minimize memory overhead.
+        """
+
+        # we recursively overwrite the model state.
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            model: TrainedModel = database.session.get(TrainedModel, model_id)
+        if not model.parent_model:
+            # base case: we can load a fully stored model.
+            policy.full_model_strategy.load_model(model_state, self._storage_dir / model.model_path)
+            return
+
+        # recursive step: we recurse to load the model state of the parent model.
+        self._reconstruct_model(model.parent_model, model_state, policy)
+
+        # we apply the incremental strategy to load our model state.
+        policy.incremental_model_strategy.load_model(model_state, self._storage_dir / model.model_path)
+
+    def _get_parent_model_id(self, pipeline_id: int, trigger_id: int) -> Optional[int]:
+        """
+        Get the id of the parent model given the trigger id of a pipeline.
+
+        Args:
+            pipeline_id: the pipeline that generated the model.
+            trigger_id: the trigger associated with the model.
+
+        Returns:
+            Optional[int]: the parent model id (if it exists).
+        """
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            previous_model: TrainedModel = (
+                database.session.query(TrainedModel)
+                .filter(TrainedModel.pipeline_id == pipeline_id, TrainedModel.trigger_id == trigger_id - 1)
+                .first()
+            )
```

Comment:
I don't know if I understand correctly. Here, we are fetching the previous model (the model from the last trigger). We need to check 3 cases:
- the previous model cannot be found (e.g., because it is deleted). Then we need to store our model fully.
- the previous model has no parent, i.e., it is stored fully. In this case, we store our model according to the incremental strategy with this model as our parent.
- the previous model has a parent (meaning that it is stored incrementally). We need to store our model also as a delta but use the parent of the preceding model as our parent.

Reply:
I think a call would be good to resolve the confusion^^
It could really be that I'm overlooking something and that there is room for optimization.